File: src/app/ball.py
from __future__ import annotations
from .player import Player


class Ball:
    def __init__(self, game: Game):
        self.game: Game = game
        self.size: float = 50
        self.x: float = 0
        self.y: float = 0
        self.speed_x: float = 0
        self.speed_y: float = 0
        self.angle: float = 100
        self.reset_pos()

    def reset_pos(self):
        self.x: float = (self.game.width - self.size) / 2
        self.y: float = (self.game.height - self.size) / 2
        self.speed_x: float = 3
        self.speed_y: float = 0

    def move(self):
        self.x += self.speed_x
        self.y += self.speed_y

    def impact_player(self, player: Player):
        self.speed_x *= -1
        self.speed_y = (self.y - (player.y + player.height / 2)) / self.angle
        if 0 < self.speed_x < 40:
            self.speed_x += 0.6
        elif self.speed_x > -40:
            self.speed_x -= 0.6

    def update_speed(self, player1: Player, player2: Player):
        if self.y + self.size >= self.game.height or self.y <= 0:
            self.speed_y *= -1
        elif self.x <= player1.x + player1.width and self.y + self.size >= player1.y and self.y < player1.y + player1.height:
            self.impact_player(player1)
        elif self.x + self.size >= player2.x and self.y + self.size >= player2.y and self.y < player2.y + player2.height:
            self.impact_player(player2)

File: src/src/ball.ts
import Game from "./game";
import Player from "./player";

export default class Ball {
  readonly angle: number = 30;
  readonly size: number = 50;

  readonly game: Game;
  x: number = 0;
  y: number = 0;
  speedX: number = 0;
  speedY: number = 0;

  constructor(game: Game) {
    this.game = game;
    this.resetPos();
  }

  resetPos() {
    this.x = (this.game.width - this.size) / 2;
    this.y = (this.game.height - this.size) / 2;
    this.speedX = 3;
    this.speedY = 0;
  }

  move() {
    this.x += this.speedX;
    this.y += this.speedY;
  }

  updateSpeed(player1: Player, player2: Player) {
    if (this.y + this.size >= this.game.height || this.y <= 0)
      this.speedY *= -1;
    else if (this.x <= player1.x + player1.width && this.y + this.size >= player1.y && this.y < player1.y + player1.height)
      this.impactPlayer(player1);
    else if (this.x + this.size >= player2.x && this.y + this.size >= player2.y && this.y < player2.y + player2.height)
      this.impactPlayer(player2);
  }

  impactPlayer(player: Player) {
    this.speedX *= -1;
    this.speedY = (this.y - (player.y + player.height) / 2) / this.angle;
    if (this.speedX > 0 && this.speedX < 40)
      this.speedX += 0.6;
    else if (this.speedX > -40)
      this.speedX -= 0.6;
  }
}
